server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1424,7 +1388,8 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                             if (endingSnapshots.contains(request.snapshot()) == false) {
                                 final SnapshotsInProgress snapshotsInProgress = newState.custom(SnapshotsInProgress.TYPE);
                                 final SnapshotsInProgress.Entry updatedEntry = snapshotsInProgress.snapshot(request.snapshot());
-                                if (updatedEntry.state().completed()) {
+                                // If the entry is still in the cluster state and is completed, try finalizing the snapshot in the repo
+                                if (updatedEntry != null && updatedEntry.state().completed()) {
```

Comment:
This was a bug introduced in #56669 , we could NPE here if a data node were to resend a state update for an already finished snapshot (master fail-over or node disconnect can easily lead to this) causing an NPE here needlessly.